library/common/http/client.cc

```diff
@@ -117,19 +164,60 @@ void Client::DirectStreamCallbacks::encodeTrailers(const ResponseTrailerMap& tra
   ASSERT(http_client_.getStream(direct_stream_.stream_handle_));
   closeStream(); // Trailers always indicate the end of the stream.
 
+  // When explicitly buffering, don't send data unless prompted.
+  if (explicit_buffering_ && bytes_to_send_ == 0) {
+    response_trailers_ = ResponseTrailerMapImpl::create();
+    HeaderMapImpl::copyFrom(*response_trailers_, trailers);
+    return;
+  }
+
+  sendTrailersToBridge(trailers);
+}
+
+void Client::DirectStreamCallbacks::sendTrailersToBridge(const ResponseTrailerMap& trailers) {
   ENVOY_LOG(debug, "[S{}] dispatching to platform response trailers for stream:\n{}",
             direct_stream_.stream_handle_, trailers);
+
   bridge_callbacks_.on_trailers(Utility::toBridgeHeaders(trailers), bridge_callbacks_.context);
   onComplete();
 }
 
-void Client::DirectStreamCallbacks::closeStream() {
-  // Envoy itself does not currently allow half-open streams where the local half is open
-  // but the remote half is closed. Note that if local is open, Envoy will reset the stream.
-  http_client_.removeStream(direct_stream_.stream_handle_);
+void Client::DirectStreamCallbacks::resumeData(int32_t bytes_to_send) {
+  ASSERT(explicit_buffering_);
+  ASSERT(bytes_to_send > 0);
+
+  bytes_to_send_ = bytes_to_send;
+
+  ENVOY_LOG(debug, "[S{}] received resume data call for {} bytes", direct_stream_.stream_handle_,
+            bytes_to_send_);
+
+  // If there is bufferd data, send up to bytes_to_send bytes.
+  // Make sure to send end stream with data only if
+  // 1) it has been received from the peer and
+  // 2) there are no trailers
+  if (hasBufferedData() || (end_stream_read_ && !end_stream_communicated_ && !response_trailers_)) {
+    sendDataToBridge(*response_data_, end_stream_read_ && !response_trailers_.get());
+    bytes_to_send_ = 0;
+  }
+
+  // If all buffered data has been sent, send and free up trailers.
+  if (!hasBufferedData() && response_trailers_.get() && bytes_to_send_ > 0) {
+    sendTrailersToBridge(*response_trailers_);
+    response_trailers_.reset();
+    bytes_to_send_ = 0;
+  }
+
+  if (deferred_error_ && bytes_to_send_ > 0) {
+    onError();
+    deferred_error_ = false;
+  }
 }
 
+void Client::DirectStreamCallbacks::closeStream() { end_stream_read_ = true; }
+
 void Client::DirectStreamCallbacks::onComplete() {
+  http_client_.removeStream(direct_stream_.stream_handle_);
```

Comment:
If you make the change above in `closeStream`, don't forget update this to remove the stream from `draining_streams_` iff `explicit_buffering_` is enabled.

Reply:
yeah this is why I actually held off - moving things around and having 2 maps felt more risky to me, but happy to give it a shot and if you two think it's cleaner the way it was we can always undo it :-)